haskell-src/Concordium/Types/Parameters.hs

```diff
@@ -27,42 +397,313 @@ import qualified Concordium.Crypto.SHA256 as Hash
+        isSupported PTTimeoutParameters ChainParametersV0 = False
+        isSupported PTTimeoutParameters ChainParametersV1 = False
+        isSupported PTTimeoutParameters ChainParametersV2 = True
+        isSupported PTMinBlockTime ChainParametersV0 = False
+        isSupported PTMinBlockTime ChainParametersV1 = False
+        isSupported PTMinBlockTime ChainParametersV2 = True
+        isSupported PTBlockEnergyLimit ChainParametersV0 = False
+        isSupported PTBlockEnergyLimit ChainParametersV1 = False
+        isSupported PTBlockEnergyLimit ChainParametersV2 = True
+        isSupported PTCooldownParametersAccessStructure cpv = supportsCooldownParametersAccessStructure (authorizationsVersionFor cpv)
+        isSupported PTFinalizationProof ChainParametersV0 = True
+        isSupported PTFinalizationProof ChainParametersV1 = True
+        isSupported PTFinalizationProof ChainParametersV2 = False
+        |]
+ )
+
+-- |Constraint on a type level 'ParameterType' that can be used to get a corresponding
+-- 'SParameterType'. An alias for 'SingI'.
+type IsParameterType (pt :: ParameterType) = SingI pt
+
+-- |Constraint on a type level 'AuthorizationsVersion' that can be used to get a corresponding
+-- 'SAuthorizationsVersion'. An alias for 'SingI'.
+type IsAuthorizationsVersion (auv :: AuthorizationsVersion) = SingI auv
+
+-- |Witness an 'IsAuthorizationsVersion' constraint using a 'SChainParametersVersion'.
+withIsAuthorizationsVersionFor :: SChainParametersVersion cpv -> (IsAuthorizationsVersion (AuthorizationsVersionFor cpv) => a) -> a
+withIsAuthorizationsVersionFor scpv = withSingI (sAuthorizationsVersionFor scpv)
+
+-- |Witness an 'IsAuthorizationsVersion' constraint using a 'SProtocolVersion'.
+withIsAuthorizationsVersionForPV :: SProtocolVersion pv -> (IsAuthorizationsVersion (AuthorizationsVersionForPV pv) => a) -> a
+withIsAuthorizationsVersionForPV spv = withSingI (sAuthorizationsVersionForPV spv)
+
+-- |A @Conditionally b a@ is an @a@ if @b ~ 'True@, and @()@ otherwise.
+data Conditionally (b :: Bool) a where
+    CFalse :: Conditionally 'False a
+    CTrue :: !a -> Conditionally 'True a
+
+instance Functor (Conditionally b) where
+    fmap _ CFalse = CFalse
+    fmap f (CTrue v) = CTrue (f v)
+
+instance Foldable (Conditionally b) where
+    foldr _ b CFalse = b
+    foldr f b (CTrue a) = f a b
+
+    foldl _ b CFalse = b
+    foldl f b (CTrue a) = f b a
+
+    foldMap _ CFalse = mempty
+    foldMap f (CTrue a) = f a
+
+instance Traversable (Conditionally b) where
+    traverse _ CFalse = pure CFalse
+    traverse f (CTrue a) = CTrue <$> f a
+
+instance (Eq a) => Eq (Conditionally b a) where
+    CFalse == CFalse = True
+    CTrue a == CTrue b = a == b
+
+instance (Ord a) => Ord (Conditionally b a) where
+    compare CFalse CFalse = EQ
+    compare (CTrue x) (CTrue y) = compare x y
+
+instance (Show a) => Show (Conditionally b a) where
+    show CFalse = "<CFalse>"
+    show (CTrue v) = show v
```

Comment:
That's not clear `(show (CTrue CFalse))`. Please add some indicator to indicate the variant.